tools/prow-auto-bumper/parser.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	return *pv.dominantVersions
+}
+
+// parse changelist, find all version changes, and store them in image name: versions map
+func (pv *PRVersions) parseChangelist(gcw *GHClientWrapper, gi gitInfo) error {
+	fs, err := gcw.ListFiles(gi.org, gi.repo, *pv.PR.Number)
+	if nil != err {
+		return err
+	}
+	for _, f := range fs {
+		if nil == f.Patch {
+			continue
+		}
+		minuses := imageMinusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, minus := range minuses {
+			iv := pv.getIndex(minus[imageImagePart], minus[imageTagPart])
+			pv.images[minus[imageImagePart]][iv].oldVersion = minus[imageTagPart]
+		}
+
+		pluses := imagePlusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, plus := range pluses {
+			iv := pv.getIndex(plus[imageImagePart], plus[imageTagPart])
+			pv.images[plus[imageImagePart]][iv].newVersion = plus[imageTagPart]
+		}
+	}
+
+	return nil
+}
+
+// Query all PRs from "k8s-ci-robot:autobump", find PR roughly 7 days old and was not reverted later.
+// Only return error if it's github related
+func getBestVersion(gcw *GHClientWrapper, gi gitInfo) (*PRVersions, error) {
+	visited := make(map[string]PRVersions)
+	var bestPv *PRVersions
+	var overallErr error
+	var bestDelta float64 = maxDelta + 1
+	PRs, err := gcw.ListPullRequests(gi.org, gi.repo, gi.getHeadRef(), gi.base)
+	if nil != err {
+		return bestPv, fmt.Errorf("failed list pull request: '%v'", err)
+	}
+
+	for _, PR := range PRs {
+		if nil == PR.State || string(ghutil.PullRequestCloseState) != *PR.State {
+			continue
+		}
+		delta := targetTime.Sub(*PR.CreatedAt).Hours()
+		if delta > maxDelta {
+			break // Over 9 days old, too old
+		}
+		pv := PRVersions{
+			images: make(map[string][]versions),
+			PR:     PR,
+		}
+		if err := pv.parseChangelist(gcw, gi); nil != err {
+			overallErr = fmt.Errorf("failed listing files from PR '%d': '%v'", *PR.Number, err)
+			break
+		}
+		vs := pv.getDominantVersions()
+		if "" == vs.oldVersion || "" == vs.newVersion {
+			log.Printf("Warning: found PR misses version change '%d'", *PR.Number)
+			continue
+		}
+		visited[vs.oldVersion] = pv
+		// check if too fresh here as need the data in visited
+		if delta < -maxDelta { // In past 5 days, too fresh
+			continue
+		}
+		if updatePR, ok := visited[vs.newVersion]; ok {
+			if updatePR.getDominantVersions().newVersion == vs.oldVersion { // The updatePR is reverting this PR
+				continue
+			}
+			if updatePR.PR.CreatedAt.Before(PR.CreatedAt.Add(time.Hour * safeDuration)) {
+				// The update PR is within 12 hours of current PR, consider unsafe
+				continue
+			}
+		}
+		if nil == bestPv || math.Abs(delta) < math.Abs(bestDelta) {
+			bestDelta = delta
+			bestPv = &pv
+		}
+	}
+	return bestPv, overallErr
+}
+
+func retryGetBestVersion(gcw *GHClientWrapper, gi gitInfo) (*PRVersions, error) {
+	var bestPv *PRVersions
+	var overallErr error
+	// retry if there is github related error
+	for retryCount := 0; nil == overallErr && retryCount < maxRetry; retryCount++ {
```

Comment:
There's a logic error here - if `nil == overallErr` you still retry until the `maxRetry` :-)

Reply:
you are right, not sure why this was not discovered in my test runs. Thanks!